cherab/solps/models/line_emitter.pyx

```diff
@@ -0,0 +1,114 @@
+    :param Plasma plasma: Plasma object. Default is None.
+    :param AtomicData atomic_data: Atomic data provider. Will use `OpenADAS` if None (default).
+    :param object lineshape: Spectral line shape model. Will use `GaussianLine` if None (default).
+    :param object lineshape_args: Positional arguments for the lineshape.
+    :param object lineshape_kwargs: Keyword arguments for the lineshape.
+    """
+
+    def __init__(self, SOLPSFunction2D radiation_function, Line line, Plasma plasma=None, AtomicData atomic_data=None,
+                 object lineshape=None, object lineshape_args=None, object lineshape_kwargs=None):
+
+        super().__init__(plasma, atomic_data)
+
+        self._line = line
+
+        self._power_density = AxisymmetricMapper(radiation_function)
+
+        self._lineshape_class = lineshape or GaussianLine
+        if not issubclass(self._lineshape_class, LineShapeModel):
+            raise TypeError("The attribute lineshape must be a subclass of LineShapeModel.")
+
+        if lineshape_args:
+            self._lineshape_args = lineshape_args
+        else:
+            self._lineshape_args = []
+        if lineshape_kwargs:
+            self._lineshape_kwargs = lineshape_kwargs
+        else:
+            self._lineshape_kwargs = {}
+
+        # ensure that cache is initialised
+        self._change()
+
+    def __repr__(self):
+        return '<SOLPSLineEmitter: element={}, charge={}, transition={}>'.format(self._line.element.name, self._line.charge, self._line.transition)
+
+    cpdef Spectrum emission(self, Point3D point, Vector3D direction, Spectrum spectrum):
+
+        cdef double ne, ni, te, radiance
+
+        # cache data on first run
+        if self._lineshape is None:
+            self._populate_cache()
+
+        # add emission line to spectrum
+        radiance = RECIP_4_PI * self._power_density(point.x, point.y, point.z)
+        return self._lineshape.add_line(radiance, point, direction, spectrum)
+
+    cdef int _populate_cache(self) except -1:
+
+        # sanity checks
+        if self._plasma is None:
+            raise RuntimeError("The emission model is not connected to a plasma object.")
+
+        if self._atomic_data is None:
+            raise RuntimeError("The emission model is not connected to an atomic data source.")
+
+        if self._line is None:
+            raise RuntimeError("The emission line has not been set.")
+
+        # locate target species
+        try:
+            self._target_species = self._plasma.composition.get(self._line.element, self._line.charge)
+        except ValueError:
+            raise RuntimeError("The plasma object does not contain the ion species for the specified line "
+                               "(element={}, ionisation={}).".format(self._line.element.symbol, self._line.charge))
+
+        # identify wavelength
+        self._wavelength = self._plasma.atomic_data.wavelength(self._line.element, self._line.charge, self._line.transition)
```

Comment:
Hi @vsnever , thanks for this PR. I have only one question. Shouldn't the wavelength be obtained from self._atomic_data instead? If the user specifies the atomic_data source, it should take the values from there in my opinion, instead from the plasma. Or you could make it optional and specify in the docstring, that if there is no atomic_data source specified, it will take the one from plasma.

Reply:
Thanks, @Mateasek, you are right. The wavelength should be always obtained from `self._atomic_data` just like in other plasma models. I'll fix this.

Reply:
Done.

Reply:
thanks